File: crates/dux-web/web/src/lib/macros.ts
// Pure helpers for the web macro surfaces. The web writes a macro's payload straight
// to the focused PTY socket rather than through a server command, so both the byte
// transform and the surface gate are mirrored from `dux_core::macros`.

import type { MacroSurface, MacroView } from "@/lib/types"
import type { SelectedTarget } from "@/lib/store"

// Builds the byte payload for a macro send, an exact port of
// `dux_core::macros::macro_payload_bytes` over UTF-8 bytes. Every newline form
// becomes Alt+Enter (ESC then CR), so a multi-line macro is entered as one prompt
// instead of submitting at each line.
export function macroPayloadBytes(text: string): Uint8Array {
  const ALT_ENTER = [0x1b, 0x0d] // ESC, CR
  const bytes = new TextEncoder().encode(text)
  const out: number[] = []
  let i = 0
  while (i < bytes.length) {
    const b = bytes[i]
    if (b === 0x0d && bytes[i + 1] === 0x0a) {
      out.push(...ALT_ENTER)
      i += 2
    } else if (b === 0x0a || b === 0x0d) {
      out.push(...ALT_ENTER)
      i += 1
    } else {
      out.push(b)
      i += 1
    }
  }
  return new Uint8Array(out)
}

// Whether a macro of `macroSurface` is available on a target of `targetKind`, an
// exact mirror of `dux_core::macros::macro_matches_surface`: "both" everywhere,
// "agent" and "terminal" only on their own target kind.
export function macroMatchesSurface(
  macroSurface: MacroSurface,
  targetKind: "agent" | "terminal",
): boolean {
  switch (macroSurface) {
    case "both":
      return true
    case "agent":
      return targetKind === "agent"
    case "terminal":
      return targetKind === "terminal"
  }
}

// The macros (in config order) available on the focused target's surface. The
// popover renders this; an empty result with a non-empty `macros` means "no
// macros for this target kind", while an empty `macros` means "none at all".
export function macrosForTarget(
  macros: MacroView[],
  target: SelectedTarget,
): MacroView[] {
  return macros.filter((m) => macroMatchesSurface(m.surface, target.kind))
}

// The surface options for the editor's Select, in config-comment order with the
// wording mirrored from the canonical `config.toml` `[macros]` comment.
export const MACRO_SURFACE_OPTIONS: {
  value: MacroSurface
  label: string
  description: string
}[] = [
  {
    value: "agent",
    label: "Agent",
    description: "Only shown when the agent pane is focused.",
  },
  {
    value: "terminal",
    label: "Terminal",
    description: "Only shown when the terminal pane is focused.",
  },
  {
    value: "both",
    label: "Both",
    description: "Shown on both surfaces.",
  },
]

// A client-side validation error for the macro editor, or null when the set is
// valid: a fast-feedback mirror of rules the server re-runs on every Save, and
// deliberately not pinned to it by a test. Drift fails safe, since a lenient client
// only earns a server refusal and a strict one only over-blocks.
export function validateMacros(macros: MacroView[]): string | null {
  const seen = new Set<string>()
  for (const macro of macros) {
    const name = macro.name.trim()
    if (name === "") return "Every macro needs a name."
    if (seen.has(name)) return `Duplicate macro name: "${name}".`
    seen.add(name)
    if (macro.text === "") return `Macro "${name}" needs some text.`
    if (!isMacroSurface(macro.surface)) {
      return `Macro "${name}" has an unknown surface.`
    }
  }
  return null
}

// Narrow an arbitrary string to a known `MacroSurface`.
export function isMacroSurface(value: string): value is MacroSurface {
  return value === "agent" || value === "terminal" || value === "both"
}

// Commit reducer for the editor's per-row form: appends when adding, otherwise
// replaces the entry at `index` in place, which is what keeps a renamed macro in its
// list position. Returns a new array and never mutates `prev`.
export function commitMacro(
  prev: MacroView[],
  index: number | "new",
  macro: MacroView,
): MacroView[] {
  if (index === "new") return [...prev, macro]
  return prev.map((m, i) => (i === index ? macro : m))
}

// Sortable ids for the editor list's drag and drop, positional rather than
// name-based: a draft may transiently hold duplicate names and the ids must be
// unique. The list changes only on drop, so they are stable for a whole drag.
export function macroDragIds(macros: MacroView[]): string[] {
  return macros.map((_, index) => `macro-${index}`)
}

// Applies a drag's end, moving the macro at `activeId`'s slot to `overId`'s.
// Returns `prev` by reference for a same-slot drop or an id that names no slot, so a
// caller can detect the no-op.
export function reorderMacrosByDrag(
  prev: MacroView[],
  activeId: string,
  overId: string,
): MacroView[] {
  if (activeId === overId) return prev
  const ids = macroDragIds(prev)
  const from = ids.indexOf(activeId)
  const to = ids.indexOf(overId)
  if (from === -1 || to === -1) return prev
  const next = prev.slice()
  const [moved] = next.splice(from, 1)
  next.splice(to, 0, moved)
  return next
}

// Single-line preview of a macro's text: newlines collapse to a visible glyph, and
// truncation counts characters rather than bytes so a multi-byte glyph never splits.
export function macroTextPreview(text: string, maxChars = 80): string {
  const oneLine = text.replace(/\r\n|\r|\n/g, " ⏎ ")
  const chars = [...oneLine]
  if (chars.length <= maxChars) return oneLine
  return chars.slice(0, maxChars).join("") + "…"
}

File: crates/dux-core/src/macros.rs
//! Shared text-macro helpers used by both surfaces.
//!
//! The byte transform ([`macro_payload_bytes`]) and the surface-filter predicate
//! ([`macro_matches_surface`]) live here — in core, not in the TUI — so a macro
//! sent from the web (via a `WireCommand`) and a macro sent from the TUI produce
//! BYTE-IDENTICAL input to the same PTY, with parity guaranteed by construction
//! rather than by keeping two copies in sync. The macro DATA model
//! ([`crate::config::MacroEntry`], [`crate::config::MacroSurface`]) stays in
//! `config.rs` next to the rest of the serde-persisted config.

use crate::config::MacroSurface;
use crate::model::SessionSurface;

/// Build the byte payload for a macro send. Newlines are translated to
/// Alt+Enter (ESC followed by CR) so that multi-line macros are entered
/// as a single multi-line prompt rather than submitting at each newline.
/// Handles `\r\n`, `\n`, and bare `\r` uniformly.
pub fn macro_payload_bytes(text: &str) -> Vec<u8> {
    const ALT_ENTER: &[u8] = b"\x1b\r";
    let bytes = text.as_bytes();
    let mut payload = Vec::with_capacity(bytes.len());
    let mut i = 0;
    while i < bytes.len() {
        match bytes[i] {
            b'\r' if bytes.get(i + 1) == Some(&b'\n') => {
                payload.extend_from_slice(ALT_ENTER);
                i += 2;
            }
            b'\n' | b'\r' => {
                payload.extend_from_slice(ALT_ENTER);
                i += 1;
            }
            b => {
                payload.push(b);
                i += 1;
            }
        }
    }
    payload
}

/// Whether a macro of `macro_surface` is available on a target of
/// `target_surface`. The pure core of the TUI's `filtered_macros` surface gate
/// (`entry.surface.matches(session_surface)`), extracted so the web's
/// `RunMacro` enforces the SAME surface restriction the TUI's macro bar does.
pub fn macro_matches_surface(macro_surface: MacroSurface, target_surface: SessionSurface) -> bool {
    macro_surface.matches(target_surface)
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn macro_payload_translates_newlines_to_alt_enter() {
        // Plain text with no newlines passes through byte-for-byte.
        assert_eq!(macro_payload_bytes("hello").as_slice(), b"hello");

        // Bare LF (\n) becomes ESC + CR.
        assert_eq!(macro_payload_bytes("a\nb").as_slice(), b"a\x1b\rb");

        // Bare CR (\r) becomes ESC + CR.
        assert_eq!(macro_payload_bytes("a\rb").as_slice(), b"a\x1b\rb");

        // CRLF collapses to a single ESC + CR (not two).
        assert_eq!(macro_payload_bytes("a\r\nb").as_slice(), b"a\x1b\rb");

        // Multiple newlines each become their own ESC + CR.
        assert_eq!(
            macro_payload_bytes("a\nb\nc").as_slice(),
            b"a\x1b\rb\x1b\rc"
        );

        // Trailing and leading newlines are also translated.
        assert_eq!(macro_payload_bytes("\na\n").as_slice(), b"\x1b\ra\x1b\r");

        // Empty input yields empty payload.
        assert!(macro_payload_bytes("").is_empty());
    }

    #[test]
    fn macro_matches_surface_mirrors_entry_matches() {
        // Both is available on every target surface.
        assert!(macro_matches_surface(
            MacroSurface::Both,
            SessionSurface::Agent
        ));
        assert!(macro_matches_surface(
            MacroSurface::Both,
            SessionSurface::Terminal
        ));
        // Agent only matches an agent target.
        assert!(macro_matches_surface(
            MacroSurface::Agent,
            SessionSurface::Agent
        ));
        assert!(!macro_matches_surface(
            MacroSurface::Agent,
            SessionSurface::Terminal
        ));
        // Terminal only matches a terminal target.
        assert!(macro_matches_surface(
            MacroSurface::Terminal,
            SessionSurface::Terminal
        ));
        assert!(!macro_matches_surface(
            MacroSurface::Terminal,
            SessionSurface::Agent
        ));
    }
}
